cache/clock_cache.cc

```diff
@@ -43,69 +46,65 @@ ClockHandleTable::~ClockHandleTable() {
+      key,
+      [&](ClockHandle* h) {
+        if (h->TryInternalRef()) {
+          if (h->IsElement() && h->Matches(key, hash)) {
+            return true;
+          }
+          h->ReleaseInternalRef();
+        }
+        return false;
+      },
+      [&](ClockHandle* h) { return h->displacements == 0; },
+      [&](ClockHandle* /*h*/) {}, probe);
+
+  // printf("Lookup: %i\n", probe);
+
+  if (e != nullptr) {
+    // TODO(Guido) Comment from #10347: Here it looks like we have three atomic
+    // updates where it would be possible to combine into one CAS (more metadata
+    // under one atomic field) or maybe two atomic updates (one arithmetic, one
+    // bitwise). Something to think about optimizing.
+    e->InternalToExternalRef();
+    e->SetHit();
+    // The handle is now referenced, so we take it out of clock.
+    ClockOff(e);
+  }
+
+  return e;
 }
 
-ClockHandle* ClockHandleTable::Insert(ClockHandle* h, ClockHandle** old) {
+ClockHandle* ClockHandleTable::Insert(ClockHandle* h, autovector<ClockHandle>* deleted, bool take_reference) {
   uint32_t probe = 0;
-  int slot = FindElementOrAvailableSlot(h->key(), h->hash, probe);
-  *old = nullptr;
-  if (slot == -1) {
-    // The key is not already present, and there's no available slot to place
-    // the new copy.
+  ClockHandle* e = FindAvailableSlot(h->key(), h->hash, probe, deleted);
+  if (e == nullptr) {
+    // No available slot to place the handle.
     return nullptr;
   }
 
-  if (!array_[slot].IsElement()) {
-    // The slot is empty or is a tombstone.
-    ClockHandle* new_entry = &array_[slot];
-    new_entry->InternalToExclusiveRef();
-    Assign(new_entry, h);
-    if (new_entry->displacements == 0) {
-      // The slot was empty.
-      return new_entry;
-    }
-    // It used to be a tombstone, so there may already be a copy of the
+  // printf("Insert: %i\n", probe);
+
+  // The slot is empty or is a tombstone. And we have an exclusive ref.
+  Assign(e, h);
+  usage_ += e->total_charge;
+  if (e->displacements != 0) {
+    // It used to be a tombstone, so there may already be copies of the
     // key in the table.
-    slot = FindElement(h->key(), h->hash, probe);
-    if (slot == -1) {
-      // Nope, no existing copy of the key.
-      return new_entry;
-    }
-    ClockHandle* old_entry = &array_[slot];
-    old_entry->ReleaseInternalRef();
-    *old = old_entry;
-    return new_entry;
-  } else {
-    // There is an existing copy of the key.
-    ClockHandle* old_entry = &array_[slot];
-    old_entry->ReleaseInternalRef();
-    *old = old_entry;
-    // Find an available slot for the new element.
-    old_entry->displacements++;
-    slot = FindAvailableSlot(h->key(), probe);
-    if (slot == -1) {
-      // No available slots.
-      return nullptr;
-    }
-    ClockHandle* new_entry = &array_[slot];
-    new_entry->InternalToExclusiveRef();
-    Assign(new_entry, h);
-    return new_entry;
+    RemoveAll(h->key(), h->hash, probe, deleted);
```

Comment:
Can't we do this outside of holding the exclusive ref?

Reply:
You may be right. I had a bad case in mind: two or more concurrent inserts (of the same key) annihilating each other. But perhaps this is not possible: the smallest insert in the probe sequence (the one whose call to FindAvailableSlot lands the smallest probe) that is not deleted during the other concurrent FindAvailableSlot calls will survive, since after this call RemoveAll only deletes larger items in the probe sequence.